samza-operator/src/main/java/org/apache/samza/operators/MessageStreamImpl.java

```diff
@@ -0,0 +1,135 @@
+ * to describe and chain the operators specs in the stream.
+ *
+ * @param <M>  type of message in this stream
+ */
+public class MessageStreamImpl<M extends Message> implements MessageStream<M> {
+
+  /**
+   * The set of operators that consume the output from this message stream.
+   */
+  private final Set<OperatorSpec> registeredOperatorSpecs = new HashSet<>();
+
+  @Override
+  public <OM extends Message> MessageStream<OM> map(MapFunction<M, OM> mapFn) {
+    OperatorSpec<OM> op = OperatorSpecs.<M, OM>createStreamOperator(m -> new ArrayList<OM>() { {
+        OM r = mapFn.apply(m);
+        if (r != null) {
+          this.add(r);
+        }
+      } });
+    this.registeredOperatorSpecs.add(op);
+    return op.getOutputStream();
+  }
+
+  @Override
+  public <OM extends Message> MessageStream<OM> flatMap(FlatMapFunction<M, OM> flatMapFn) {
+    OperatorSpec<OM> op = OperatorSpecs.createStreamOperator(flatMapFn);
+    this.registeredOperatorSpecs.add(op);
+    return op.getOutputStream();
+  }
+
+  @Override
+  public MessageStream<M> filter(FilterFunction<M> filterFn) {
+    OperatorSpec<M> op = OperatorSpecs.<M, M>createStreamOperator(t -> new ArrayList<M>() { {
+        if (filterFn.apply(t)) {
+          this.add(t);
+        }
+      } });
+    this.registeredOperatorSpecs.add(op);
+    return op.getOutputStream();
+  }
+
+  @Override
+  public void sink(SinkFunction<M> sinkFn) {
+    this.registeredOperatorSpecs.add(OperatorSpecs.createSinkOperator(sinkFn));
+  }
+
+  @Override
+  public <WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> MessageStream<WM> window(
+      Window<M, WK, WV, WM> window) {
+    OperatorSpec<WM> wndOp = OperatorSpecs.createWindowOperator((WindowFn<M, WK, WS, WM>) window.getInternalWindowFn());
+    this.registeredOperatorSpecs.add(wndOp);
+    return wndOp.getOutputStream();
+  }
+
+  @Override
+  public <K, JM extends Message<K, ?>, RM extends Message> MessageStream<RM> join(MessageStream<JM> otherStream,
+      JoinFunction<M, JM, RM> joinFn) {
+    MessageStreamImpl<RM> outputStream = new MessageStreamImpl<>();
+
+    BiFunction<M, JM, RM> parJoin1 = joinFn::apply;
+    BiFunction<JM, M, RM> parJoin2 = (m, t1) -> joinFn.apply(t1, m);
+
+    // TODO: need to add default store functions for the two partial join functions
+
+    ((MessageStreamImpl<JM>) otherStream).getRegisteredOperatorSpecs().add(
+        OperatorSpecs.<JM, K, M, RM>createPartialJoinOperator(parJoin2, outputStream));
+    this.getRegisteredOperatorSpecs().add(OperatorSpecs.<M, K, JM, RM>createPartialJoinOperator(parJoin1, outputStream));
+    return outputStream;
+  }
+
+  @Override
+  public MessageStream<M> merge(Collection<MessageStream<M>> otherStreams) {
+    MessageStreamImpl<M> outputStream = new MessageStreamImpl<>();
+
+    otherStreams.add(this);
+    otherStreams.forEach(other ->
+        ((MessageStreamImpl<M>) other)
+            .getRegisteredOperatorSpecs()
```

Comment:
Yeah that'll be better. I'll make these changes:
1. Mutate the registeredOperatorSpecs set directly within MessagesImpl.
2. Make getRegisteredOperatorSpecs() return an immutable collection.